File: sudoku_dynamic/grid.py
import numpy as np

class Grid:
    def __init__(self, grid_size=9, box_size=3):
        self.GRID_SIZE = grid_size
        self.BOX_SIZE = box_size
        self.grid = np.zeros((grid_size, grid_size)).astype(int)

    def getRow(self, row):
        return self.grid[row]
    
    def getCol(self, col):
        return self.grid[:,col].reshape(1,-1)
    
    def getBox(self, row, col):
        org_row = row // self.BOX_SIZE
        org_col = col // self.BOX_SIZE
        return self.grid[org_row:org_row+self.BOX_SIZE, org_col:org_col+self.BOX_SIZE].reshape(1,-1)

File: sudoku_static/grid.cpp
#include "grid.h"
#include<cstring>
using namespace std;

Grid::Grid(int grid_size, int box_size){
	GRID_SIZE = grid_size;
	BOX_SIZE = box_size;
	grid = new char[GRID_SIZE*GRID_SIZE];
}

char* Grid::getRow(int row, char* data){
	for (int i=0; i<GRID_SIZE; i++){
		data[i] = grid[row*GRID_SIZE+i];
	}
	
	return data;
}

char* Grid::getCol(int col, char* data){
	for (int i=0; i<GRID_SIZE; i++){
		data[i] = grid[i*GRID_SIZE+col];
	}
	return data;
}	

char* Grid::getBox(int row, int col, char* data){
	int org_row = row/BOX_SIZE; //box起始行
	int org_col = col/BOX_SIZE; //box起始列
	int idx = 0; 
	for (int r=org_row; r<org_row+BOX_SIZE; r++){
		for (int c=org_col; c<org_col+BOX_SIZE; c++){
			data[idx] = grid[r*GRID_SIZE+c];
			idx++;
		}
	}
	return data;
}

Grid::~Grid(){
	delete[] grid;
}
